rasa/utils/tensorflow/model_data.py

```diff
@@ -24,21 +24,85 @@
 logger = logging.getLogger(__name__)
 
 
-# Mapping of attribute name and feature name to a list of numpy arrays representing
-# the actual features
-# For example:
-# "text" -> { "sentence": [
-#   "numpy array containing dense features for every training example",
-#   "numpy array containing sparse features for every training example"
-# ]}
-Data = Dict[Text, Dict[Text, List[np.ndarray]]]
+class FeatureArray(np.ndarray):
+    # Subclassing np.array: https://numpy.org/doc/stable/user/basics.subclassing.html
+
+    def __new__(
+        cls, input_array: np.ndarray, number_of_dimensions: int,
+    ):
+        feature_array = np.asarray(input_array).view(cls)
+
+        if number_of_dimensions <= 2:
+            feature_array.feature_dimension = input_array.shape[-1]
+            feature_array.is_sparse = isinstance(input_array[0], scipy.sparse.spmatrix)
+        elif number_of_dimensions == 3:
+            feature_array.feature_dimension = input_array[0].shape[-1]
+            feature_array.is_sparse = isinstance(input_array[0], scipy.sparse.spmatrix)
+        else:
+            feature_array.feature_dimension = input_array[0][0].shape[-1]
+            feature_array.is_sparse = isinstance(
+                input_array[0][0], scipy.sparse.spmatrix
+            )
+
+        feature_array.number_of_dimensions = number_of_dimensions
+
+        return feature_array
+
+    def __array_finalize__(self, obj):
+        if obj is None:
+            return
+
+        self.feature_dimension = getattr(obj, "feature_dimension", None)
+        self.number_of_dimensions = getattr(obj, "number_of_dimensions", None)
+        self.is_sparse = getattr(obj, "is_sparse", None)
+
+        default_attributes = {
+            "feature_dimension": self.feature_dimension,
+            "number_of_dimensions": self.number_of_dimensions,
```

Comment:
I think `feature_dimension` is a confusing name (since features have multiple dimensions). Not sure what would be better name. In keras this number is called `units`.